### ISSUE TITLE
Severe slowdown with 1 BN step in while_v2

### ISSUE THREAD

**Issue by OverLordGoldDragon** [tensorflow/tensorflow#36797]
I'm developing recurrent batch normalization for LSTM, and am seeing a best-case scenario of **8x slowdown** relative to base implementation (worst-case can **exceed 20x**). `tf.function` helps notably, but not nearly enough - and the behavior is unreasonable; applying the operation on _1 out of 150 steps_ yields a 6.7x slowdown, and on 150/150 steps, 8x. I suspect [`while_v2.while_loop()`](https://github.com/tensorflow/tensorflow/blob/r2.1/tensorflow/python/ops/control_flow_ops.py#L2666)'s gradient handling is responsible.

The problem's exacerbated by TF 2.1's [broken profiler](https://github.com/tensorflow/tensorboard/issues/3256) - my IDE (Spyder 4) is also unable to capture the call details in Eager or Graph, unlike in TF2. I also cannot develop in 2.0 per bug fixes and significant performance improvements. I don't think `while_v2` was designed to handle whatever I'm doing.

What can I do to circumvent such a severe slowdown? Also, is this the best place to ask, or is there a dev forum somewhere to avoid a month-long back & forth? Help is appreciated.

<hr>

**DETAILS**: a part of the full code:

```python
def recurrent_bn(self, inputs_t, op_name, step, training=None):
    training = self._get_training_value(training)
    beta, gamma, moving_mean, moving_variance = self._get_bn_vars(op_name)

    @tf.function
    def outs(inputs_t, step, beta, gamma, moving_mean, moving_variance, training):
        if tf.math.less(step, self.max_inference_step) and (training != False):
            mean_t, variance_t = self._get_stats_and_maybe_update(
                inputs_t, step, moving_mean, moving_variance, training)
        else:
            mean_t, variance_t = moving_mean[-1], moving_variance[-1]
        return nn.batch_normalization(inputs_t, mean_t, variance_t,
                                      beta, gamma, self.bn_epsilon)

    outputs = outs(inputs_t, step, beta, gamma, moving_mean,
                   moving_variance, training)
    outputs.set_shape(inputs_t.shape)
    return outputs

def _get_bn_vars(self, op_name):
    return [getattr(self, op_name + '_' + var_name) for var_name in
            ('gamma', 'beta', 'moving_mean', 'moving_variance')]  # these were built
```

`recurrent_bn` is applied 9 times in `call` - 4 times for each of `i, f, c, o` of `kernel, recurrent`, and once on `c`. `_get_stats_and_maybe_update` returns `nn.moments` applied on `moving_...[step:step+1]` slice, and calls [`_assign_moving_average`](https://github.com/tensorflow/tensorflow/blob/r2.1/tensorflow/python/keras/layers/normalization.py#L503) (added method to `LSTMCell`) when `training` is truthy. `LSTM` and `LSTMCell` are also inherited by `recurrent_v2` classes, which yielded a speedup.

 - `gamma, beta` have shape `(1, self.units)`
 - `moving_mean, moving_variance` have shape `(self.max_inference_steps, self.units)`
 - There are 9 of each -> 36 total
 - Example call: `x_i = self.recurrent_bn(x_i, 'kernel_i', step, training)`

<hr>

**Possible culprits**:

 1. Gradients, **main culprit**, via T2-T5: T3 yielded significant speedup relative to every other test, including T4 - and T5 was negligible. It appears that updating the BN variables just _once_ triggers some gradient operations for _every timestep_ - hence the slowdown. 
 2. `_get_bn_vars`, **unknown**; the variables are being fetched from a Python dictionary in `tf.function`, and though there are no retracing warnings, it's still a suspect
 3. `nn.batch_normalization` + `nn.moments`, **ruled out** via T2; only a 12% speedup is granted.
 4. `step`, **ruled out** via T6 + T7; no noted performance difference
 5. `CuDNNLSTM`, **ruled out**; I set `implementation = 1` and `recurrent_dropout = .05`

**Tests**:

 - **T1**: replace `self.max_inference_steps` (`.numpy() == 150`) with `tf.constant(1, dtype='int64')
 - **T2**: T1 + move `nn.batch_normalization` under `if`, replace `else` with `return inputs_t`
 - **T3**: wrap `gamma, beta` with `K.stop_gradient()` _before_ `if`, and `mean_t, variance_t` _after_ `else` (otherwise cannot slice wrapped tensor)
 - **T4**: T3, but wrap everything _after_ `else`.
 - **T5**: `K.stop_gradient` on `inputs_t, training`
 - **T6**: immediately `return inputs_t` in `recurrent_bn`
 - **T7**: remove hidden state `step`, don't return `step + 1` in `call`

**Bugs**:

  1. Wrapping `self._assign_moving_average` with `self.add_update` fails [`_assert_same_graph()`](https://github.com/tensorflow/tensorflow/blob/r2.1/tensorflow/python/framework/ops.py#L5816):

```
original_item == updates[0], item == updates[1]
updates[0] == <tf.Operation 'training/Adam/Adam/AssignAddVariableOp' type=AssignAddVariableOp>
updates[1] == <tf.Operation 'AssignMovingAvg' type=ResourceStridedSliceAssign>
updates[0].graph == <tensorflow.python.framework.ops.Graph object at 0x000002A3473AF788>
updates[1].graph == <tensorflow.python.ops.control_flow_v2_func_graphs.CondBranchFuncGraph
                     object at 0x000002A35C7F2548>
```
   -  [`updates`](https://github.com/tensorflow/tensorflow/blob/r2.1/tensorflow/python/keras/engine/training.py#L2059); `len(updates) == 37` = 9*4 + 1; so 4 per each of 9 calls to `recurrent_bn`, +1 default
   - [`self.add_update`](https://github.com/tensorflow/tensorflow/blob/r2.1/tensorflow/python/keras/layers/recurrent_v2.py#L1158) is used in `LSTM` right _after_ `K.rnn` - so it appears that `while_v2` is responsible for `updates[1].graph` being different
   - `self.add_update` compiles, and yields a 10% speedup in Eager

 2. As noted [here](https://github.com/tensorflow/tensorflow/issues/36840): while there aren't errors in Eager, the gradients are extremely small: `1e-9` to `1e-19`, whereas usually the same layers have `1e-6` to `1e-2`. This may or may not be a design rather than a bug problem.

<hr>

**IDE Profiler encapsulation**:

<img src="https://user-images.githubusercontent.com/16495490/74617169-49255c80-5145-11ea-9aba-b5cca922e97c.png" width="550">

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
